python/mead/downloader.py

```diff
@@ -0,0 +1,300 @@
+        zip_ref.extractall(temp_file)
+    return temp_file
+
+
+def mime_type(loc):
+    import magic
+    return magic.Magic(mime=True).from_file(loc)
+
+
+def get_cache_dir():
+    import json
+    try:
+        data_cache = json.load(open("config.json")).get("datacache", "~/.bl-dataset-embeddings/")
+    except IOError:
+        data_cache = "~/.bl-dataset-embeddings/"
+
+    return data_cache
+
+
+def extractor(filepath, cache_dir, extractor_func):
+    import hashlib
+    import shutil
+    import os
+    sha1 = hashlib.sha1(open(filepath, 'rb').read()).hexdigest()
+    print("extracting file..")
+    path_to_save = filepath if extractor_func is None else extractor_func(filepath)
+    if not os.path.exists(cache_dir):
+        os.makedirs(cache_dir)
+    path_to_save_sha1 = os.path.join(cache_dir, sha1)
+    shutil.move(path_to_save, path_to_save_sha1)
+    print("downloaded data saved in {}".format(path_to_save_sha1))
+    return path_to_save_sha1
+
+
+def web_downloader(url):
+    import requests
+    import shutil
+    from baseline.progress import create_progress_bar
+    import os
+    r = requests.get(url, stream=True)
+    if r.status_code != 200:
+        raise RuntimeError("The file can not be downloaded")
+    path_to_save = "/tmp/data.dload-{}".format(os.getpid())
+    try:
+        print("downloading {}".format(url))
+        total_length = None
+        with open(path_to_save, 'wb') as f:
+            total_length = r.headers.get('content-length', None)
+            if total_length is not None:
+                total_length = int(total_length)
+                chunk_size = 1024
+                pg = create_progress_bar(total_length/chunk_size)
+                for chunk in r.iter_content(chunk_size=chunk_size):
+                    if chunk:
+                        f.write(chunk)
+                        pg.update()
+                        f.flush()
+                pg.done()
+            else:
+                raise RuntimeWarning("Total length can not be calculated, something wrong with the download file, trying ")
+                shutil.copyfileobj(r.raw, f)
+
+    except:  # this is too broad but there are too many exceptions to handle separately
+        raise RuntimeError("failed to download data from [url]: {} [to]: {}".format(url, path_to_save))
+    return path_to_save
+
+
+def validate_url(url):
+    import re
+    regex = re.compile(
+        r'^(?:http|ftp)s?://' # http:// or https://
+        r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|' #domain...
+        r'localhost|' #localhost...
+        r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})' # ...or ip
+        r'(?::\d+)?' # optional port
+        r'(?:/?|[/?]\S+)$', re.IGNORECASE)
+    return re.match(regex, url) is not None
+
+
+def update_cache(key):
+    import json
+    try:
+        dcaches = json.load(open(os.path.join(self.data_download_cache, "datasets-embeddings-cache.json")))[key]
```

Comment:
This should probably be a shorter name, also saved in a global or something like `CACHE = datasets-embeddings-cache.json`